src/StatusBar.ts

```diff
@@ -0,0 +1,86 @@
+import { window, StatusBarAlignment } from 'vscode'
+import * as elegantSpinner from 'elegant-spinner'
+
+type Status = 'running' | 'failed' | 'success' | 'stopped' | 'initial'
+type QueueItem = {
+  source: string
+  status: Status
+  details: string | undefined
+}
+
+// The bottom status bar
+export class StatusBar {
+  private statusBarSpinner: NodeJS.Timer
+  private statusBarItem = window.createStatusBarItem(StatusBarAlignment.Left)
+  private statusKey = 'Jest:'
+  private frame = elegantSpinner()
+  private priorities: Status[] = ['running', 'failed', 'success', 'stopped', 'initial']
+  private queue: QueueItem[] = []

```

Comment:
what do users expect from status bar? 
1. the overall project status 
2. the current workspace status. 

For a single workspace project, these status are the same; for multi-workspace, they are different. The current implementation reported what it considered the most _important_ status (by type), which doesn't guarantee to answer either question above. The good news is I think it can be easily addressed without much structure change outside of this file.

1. instead of having a single queue, we can have a dictionary: `{[source]: QueueItem}`,  which holds the latest status of each workspace. 
2. this status dictionary can be used to report project-wide status, such as `running:1, success:3, fail:2` etc, to save space, we should try to replace the text with icon. 
3. this status dictionary can also be used to report "current" workspace status, in the bind() method, you can add a new `setActive()` for example, to display the current workspace's status, in addition to the project-wide status above. `statusBar.setActive()` can be invoked in `JestExt.DidChangeActiveTextEditor()`, for example. So the final status could look like: `Jest: current: 1 test suite failed (all: success:2, fail:1)`...
4. we could even consider creating 2 `statusBarItem` for active workspace and project-wide status, each is associated with its own `statusBarCommand` showing corresponding output channel. The project-level channel would simply report the detail of the whole status dictionary.

This should simplify the logic (no need for priority, queuing etc) a bit while providing more relevant and complete information, compatible with the current implementation. what do you think?

Reply:
@connectdotz  Is this really a block for merging this PR? This sounds more like improvement because @escaton solution works. I personally never look at status bar for jest extension status, however I'm using `debug` single test feature a lot which currently doesn't work at all in multi root. Can we merge this and create a new ticket with your suggestions?

Reply:
I believe that displaying
```
Jest: current: 1 test suite failed (all: success:2, fail:1)
```
is far from good UX. There is not much space in status bar to give it's 1/4 (on my mbp) to one extension.
And also
>Why are we so sure that a failure status from other workspace is more important than the workspace users are currently working on?

Why should we separate code to workspaces in this case? It's all one project with one common status. Your changes in one workspace could potentially broke tests in another workspace. Why should we still message you "Success" in the first place?

Anyway i'm +1 to discuss it in separate issue.

Reply:
sorry for the delay, just returned from a long overseas trip. 

@escaton I see where you come from... you are saying that regardless of single or multiple workspaces, if they are under the same project, we should just display a single status; but given each workspace does report individual status, you had to "hand pick" one thus the priority: `'running' >  'failed' > 'success' > 'stopped' > 'initial'`. 

> Your changes in one workspace could potentially broke tests in another workspace

True, but the other way could also be true: imagine you are working on your own workspace X, but the project shows failure status from other workspace Y that you know has nothing to do with your change. Because the status bar will always show failure from Y, You will not know if your change has caused new test failures (in X) or corrected existing test failures (in X)...  In such a scenario, you can probably understand why the status bar is either useless or confusing, especially when other users are not as familiar with the priority logic as you...

As a tool builder, we are humbled by vast use cases our users presented. Therefore, we always strive to provide a solution with as few assumptions as possible so they can work with more use cases.

> ...is far from good UX. There is not much space in status bar to give its 1/4 (on my mbp) to one extension.

probably, then let's see how we can improve the UX, without compromising the principle we discussed above 😄 ...

Reply:
> but the project shows failure status from other workspace Y that you know has nothing to do with your change. Because the status bar will always show failure from Y, You will not know if your change has caused new test failures (in X) or corrected existing test failures (in X)

This sounds like you may want to disable extension for that workspace. Since you don't know what to do with that failures.

Reply:
> This sounds like you may want to disable extension for that workspace. Since you don't know what to do with those failures.

you could, but wouldn't that be hacky that you will have to update the workspaces settings based on the change you are working on? What if you also need to work on workspace Z, which does impact Y,  you will have to switch the settings again?

Reply:
no, i don't mean you should change project settings once you switch to other code. I mean that if you can't handle errors in Y than maybe you want to mute them all by disabling jest in Y forever.

Reply:
> if you can't handle errors in Y than maybe you want to mute them all by disabling jest in Y forever.

But you can't disable Y **forever** because sometimes you need to work on Z, which could impact Y; therefore Y should not be disabled in that situation, right? You can probably see the conflict and overhead this approach could impose when working on a larger project or multiple folders... 

Don't give up on your original argument for the importance of the whole project status; it is not wrong, just incomplete for some use cases -- We probably do not want to exclude Y because it is indeed part of the project that we ultimately care. 

The examples above merely points out sometimes we do need to know the status of the workspace one is working on, **in addition to** the project level status. If we allow ourselves to take a step back and question our assumption: why single status? What not display both? Wouldn't that support more use cases with less assumption? "But the status bar has limited space!" sure, so the problem changed from **"how do we choose one status for all use cases"**  to **"how to display both statuses in limited space"** , a much simpler problem that I am confident we will find a satisfying solution quickly.

I understand you might not have time to address this issue, that is quite all right. Once we merge this PR into a branch, others can start to contribute. You have established a great starting point, and we will get this released soon.